src/mon/PaxosService.cc

```diff
@@ -385,20 +385,38 @@ void PaxosService::maybe_trim()
   }
 
   version_t to_remove = trim_to - get_first_committed();
-  if (g_conf()->paxos_service_trim_min > 0 &&
-      to_remove < (version_t)g_conf()->paxos_service_trim_min) {
+  version_t paxos_service_trim_min = g_conf().get_val<version_t>("paxos_service_trim_min");
+  version_t paxos_service_trim_max = g_conf().get_val<version_t>("paxos_service_trim_max");
+  uint64_t paxos_service_trim_max_multiplier = g_conf().get_val<uint64_t>("paxos_service_trim_max_multiplier");
+
+  if (paxos_service_trim_min > 0 &&
+      to_remove < paxos_service_trim_min) {
     dout(10) << __func__ << " trim_to " << trim_to << " would only trim " << to_remove
-	     << " < paxos_service_trim_min " << g_conf()->paxos_service_trim_min << dendl;
+	     << " < paxos_service_trim_min " << paxos_service_trim_min << dendl;
     return;
   }
 
-  if (g_conf()->paxos_service_trim_max > 0 &&
-      to_remove > (version_t)g_conf()->paxos_service_trim_max) {
-    dout(10) << __func__ << " trim_to " << trim_to << " would only trim " << to_remove
-	     << " > paxos_service_trim_max, limiting to " << g_conf()->paxos_service_trim_max
+  if (paxos_service_trim_max > 0 &&
+      to_remove > paxos_service_trim_max ) {
+    version_t trim_difference = to_remove - paxos_service_trim_max;
+    if (trim_difference > paxos_service_trim_max * 0.5) {
```

Comment:
the policy is encoded with both the settings (numbers) and the code here. i'd suggest make the magic numbers and the policy more explicit if we have to encode them:

```c++
// cap the number of states to trim
to_remove = [to_remove, this] {
  const version_t trim_max =
    g_conf().get_val<version_t>("paxos_service_trim_max");
  if (trim_max == 0 || to_remove < trim_max) {
    // do not cap it
    return to_remove;
  }
  if (to_remove < trim_max * 1.5) { /// XXX: magic number here
    // cap it with a soft limit
    return trim_max;
  }
  // increase trim_max adaptively if to_remove is too large
  const version_t new_trim_max = (trim_max + to_remove) / 2;
  // but set an upper bound for it.
  const uint64_t trim_max_multiplier = 
    g_conf().get_val<uint64_t>("paxos_service_trim_max_multiplier");
  if (trim_max_multiplier) {
    return std::min(new_trim_max, trim_max * trim_max_multiplier);
  } else {
    return new_trim_max;
  }
}();
trim_to = get_first_committed() + to_remove;
```

Reply:
Okay sure I'll make the change.